### ISSUE TITLE
RFC/ENH?: `optimize.curve_fit`: option to use global optimization for curve fit

### ISSUE THREAD

**Issue by mdhaber** [scipy/scipy#21102]
### Is your feature request related to a problem? Please describe.

In gh-16191, it was suggested that `curve_fit` might benefit from an option to use global optimization. 

### Describe the solution you'd like.

Based on drafting the code, I think this would be relatively easy to do, and it would fit into existing interface nicely. All the parameters would work with the exception that not all optimizers accept/use `p0`, `bounds`, `jac`.

The method names could be those of global optimizers, and there is already `kwargs` for passing additional optimizer options through (e.g. an RNG for `differential_evolution`). `pcov` could be estimated using `scipy.differentiate.hessian` (gh-20708). (Some underlying local optimizers return a Hessian estimate, but presumably if the user is willing to use global optimization, speed is a secondary concern.)

If desired, I could deprecate `full_output` at the same time or separately, instead packing all the outputs into a `make_tuple_bunch` with `pop` and `pcov` unpackable and the rest accessible (only) by attribute. If there is anything else about the interface of `curve_fit` that you'd like to see improved at the same time (e.g gh-18925), I can look into the feasibility of that, too.

### Describe alternatives you've considered.

- Adding this feature at the level of `scipy.optimize.least_squares`. That would be quite a bit more work because there is much more to the interface. Seems like that could be done later (*maybe* reworking `curve_fit` to use it, if it makes it any simpler) if desired.
- Making a decision against such an addition.

### Additional context (e.g. screenshots, GIFs)

Thoughts? Ping @andyfaff @AtsushiSakai @tupui @dschmitz89 @maurosilber @GalagaBits since it looks like you've contributed toward `curve_fit` in recent years.

**Comment by andyfaff** [comment]
There are several factors to consider here. 

First of all, there are various packages that augment the various minimisers and least squares algorithms for curvefitting purposes. For example [lmfit](https://lmfit.github.io/lmfit-py/). Functionality that we introduce will be increasing the overlap, how far do we go down that path.

There is a large list of what would be useful, and fulfilling that list would be a big workload.

Just off the top of my head:

- why not use any `minimize` method, as well as the global optimisers?
- for each minimisation method we use we might want to offer all their respective options.
- it would be nice to use `NonLinearConstraint, LinearConstraint` for implementing constraints. How does their use interact with e.g. calculating covariance matrices?
- it would be nice to hold and release different entries without having to reparameterise the system.
- ability to change fit metric. For example, with scalar minimisation we could use negative log-posterior (i.e. include the prior terms in the statistic).

I'm not sure that fits in a single function.

**Comment by andyfaff** [comment]
> Can't that be done (theoretically) with setting the lower and upper bounds equal? Maybe a subclass of Bounds for that purpose would do the trick.

At a higher level it would be easier to set a mask, e.g.
`varying = [True, False, False, True]` allows you to just vary parameters 0 and 3. Adjusting a Bounds class for this is more heavyweight (and is not currently implemented for least_squares or leastsq. Setting ub=lb is probably functionally equivalent, not sure if there are edge cases.

**Comment by mdhaber** [comment]
Hah heavyweight in a function that's doing global optimization? : )

All I'm after, though, is that the `bounds` input could be used for this if we don't want to increase the footprint. 

And internally (in `minimize`, at least) I think that we already look for equal lower and upper bounds and filter them out of the problem, right? We did that just for this purpose, IIRC. I'll check later.

**Comment by andyfaff** [comment]
Yes, we filter in `minimize`. I can't remember how it works when one has constraints.

**Comment by mdhaber** [comment]
Yeah we remove fixed variables from the constraints, too.
https://github.com/scipy/scipy/blob/87c46641a8b3b5b47b81de44c07b840468f7ebe7/scipy/optimize/_minimize.py#L690

And because the variable is fixed, we NaN the corresponding values of the Jacobian.

We only do this for a few methods, and typically only if they need finite difference derivatives. I guess the others do OK without removing the variables, so we left them alone. I suppose we can test the global optimizers to see how they do.

**Comment by tupui** [comment]
Just a fly by comment. This function is consistently the most consulted in our doc. Difficult to say why of course, but at least something to consider.

**Comment by dschmitz89** [comment]
> Just a fly by comment. This function is consistently the most consulted in our doc. Difficult to say why of course, but at least something to consider.

Can relate to that, `curve_fit` was also the first scipy function I ever used.

I tend to agree with @andyfaff here. While adding support for more optimizers is not super difficult, it will become an API nightmare. Global optimizers need bounds, not a starting point `p0` to begin with.

And we would effectively create a less feature complete copy of lmfit. lmfit contains all kinds of niceties such as keeping parameters constant, more sophisticated uncertainty calculations via F test (or even MCMC via `emcee`) and a more convenient interface. What about adding a link to lmfit in the docs? In the stats docs, we also link to PyMC, statsmodels etc.

**Comment by mdhaber** [comment]
> While adding support for more optimizers is not super difficult, it will become an API nightmare. Global optimizers need bounds, not a starting point p0 to begin with.

Just thought I'd mention that `p0` and `bounds` are both existing parameters. It seems the addition to the documentation would be about as complex as listing the methods that use `p0` in the `p0` description and similarly in `bounds`, like how `minimize` documents the methods that use `bounds`/`constraints`. Is that the concern, or would it be worse than that?

> What about adding a link to lmfit in the docs? In the stats docs, we also link to PyMC, statsmodels etc.

This is fine. Does that mean that we would consider `curve_fit` to be in maintainance-only mode?

**Comment by dschmitz89** [comment]
> > While adding support for more optimizers is not super difficult, it will become an API nightmare. Global optimizers need bounds, not a starting point p0 to begin with.
> 
> Just thought I'd mention that `p0` and `bounds` are both existing parameters. It seems the addition to the documentation would be about as complex as listing the methods that use `p0` in the `p0` description and similarly in `bounds`, like how `minimize` documents the methods that use `bounds`/`constraints`. Is that the concern, or would it be worse than that?
> 

That is the concern yes. 

> > What about adding a link to lmfit in the docs? In the stats docs, we also link to PyMC, statsmodels etc.
> 
> This is fine. Does that mean that we would consider `curve_fit` to be in maintainance-only mode?

I would not consider it maintenancy only but not add functionality that is available in other popular packages. Adding the `nan_policy` was very useful for example. `lmfit` very successfully fills the missing bits in `curve_fit` concerning the usage of other optimizers in my opinion. And the overall churn of passing through arguments in the different `optimize` hierarchies has caused a lot of trouble already, at least in my opinion.

**Comment by mdhaber** [comment]
> I would not consider it maintenancy only but not add functionality that is available in other popular packages. Adding the nan_policy was very useful for example.

Hmm wouldn't that make us tend toward adding less popular features that aren't found elsewhere? I would have thought the opposite - SciPy should have the features that most users would most commonly expect, and users can turn to more specialized package for more specific needs.

Anyway, I'm not pushing either way for this feature, personally. Just trying to explore it and push for a resolution either way. Thanks for your thoughts.

**Comment by mdhaber** [comment]
This is roughly what we'd be looking at (before adding tests).
https://github.com/scipy/scipy/compare/main...mdhaber:gh21102?expand=1

It allows for all methods of `minimize` and all global optimizers. It can pass them their respective options (collected by existing `**kwargs` parameter of `curve_fit`). `p0`, `bound`, and/or `jac` are used if they are supported by the underlying optimizer. The inverse of the covariance matrix is estimated using `differentiate.hessian` if `jac` is not a callable and `differentiate.jacobian` (jacobian of the gradient) otherwise. I think all other options of `curve_fit` (e.g. `sigma`, `absolute_sigma`, `check_finite`, `nan_policy`) should "just work" because they are dealt with before we get to the choice of method.

It gives a sense of how the documentation would end up, although it's not polished.

LMK if this looks promising and I'll add tests, or let me know if I should let it go, and we'll close this. (That's fine, too.)

**Comment by ev-br** [comment]
(coming here after an offline discussion with Matt)

FWIW, I second the points @andyfaff makes in https://github.com/scipy/scipy/issues/21102#issuecomment-2204778041

Currently, `curve_fit` is a convenience wrapper on top of local LSQ minimizers with presumably more "user-friendly" UI. 
And I do believe that global optimization is a different enough problem to warrant keeping it separate from local optimization.
